finality-aleph/src/new_network/connection_manager/service.rs

```diff
@@ -0,0 +1,662 @@
+            }) => {
+                let (addresses, responses) = discovery.handle_message(message, handler);
+                let maybe_command = match addresses.is_empty() {
+                    false => {
+                        self.connections
+                            .add_peers(session_id, addresses.iter().flat_map(get_peer_id));
+                        Some(ConnectionCommand::AddReserved(
+                            addresses.into_iter().map(|address| address.0).collect(),
+                        ))
+                    }
+                    true => None,
+                };
+                (
+                    maybe_command,
+                    responses.into_iter().map(Self::network_message).collect(),
+                )
+            }
+            None => {
+                debug!(target: "aleph-network", "Received message from unknown session: {:?}", message);
+                (None, Vec::new())
+            }
+        }
+    }
+
+    /// Returns a data sink for data received that is associated with the given session.
+    pub fn get_session_channel(&self, session_id: &SessionId) -> Option<&mpsc::UnboundedSender<D>> {
+        self.sessions
+            .get(session_id)
+            .map(|session| session.data_for_user.as_ref())
+            .flatten()
+    }
+}
+
+/// Input/output interface for the connectiona manager service.
+pub struct IO<D: Clone + Codec> {
+    commands_for_network: mpsc::UnboundedSender<ConnectionCommand>,
+    messages_for_network: mpsc::UnboundedSender<(NetworkData<D>, DataCommand)>,
+    commands_from_user: mpsc::UnboundedReceiver<SessionCommand<D>>,
+    messages_from_user: mpsc::UnboundedReceiver<(D, SessionId, Recipient)>,
+    messages_from_network: mpsc::UnboundedReceiver<NetworkData<D>>,
+}
+
+#[derive(Debug)]
+enum SendError {
+    Network,
+    Command,
+    User,
+}
+
+impl<D: Clone + Codec> IO<D> {
+    fn send_data(&self, to_send: (NetworkData<D>, DataCommand)) -> Result<(), SendError> {
+        self.messages_for_network
+            .unbounded_send(to_send)
+            .map_err(|_| SendError::Network)
+    }
+
+    fn send_command(&self, to_send: ConnectionCommand) -> Result<(), SendError> {
+        self.commands_for_network
+            .unbounded_send(to_send)
+            .map_err(|_| SendError::Command)
+    }
+
+    fn send(
+        &self,
+        to_send: (
+            Option<ConnectionCommand>,
+            Vec<(NetworkData<D>, DataCommand)>,
+        ),
+    ) -> Result<(), SendError> {
+        if let Some(command) = to_send.0 {
+            self.send_command(command)?;
+        }
+        for data_to_send in to_send.1 {
+            self.send_data(data_to_send)?;
+        }
+        Ok(())
+    }
+
+    fn on_data<NI: NetworkIdentity>(
+        &self,
+        service: &Service<NI, D>,
+        data: D,
+        session_id: SessionId,
+    ) -> Result<(), SendError> {
+        if let Some(data_for_session_user) = service.get_session_channel(&session_id) {
```

Comment:
Ok I see why ref. Maybe instead of function `get_session_channel` we can have function `fn send_data_to_session(session_id, msg) -> Result<...>`. Wdyt?

Reply:
Oh, yeah, I guess that makes sense, although it will be slightly less convenient to test.

Reply:
Changed, see below.